File: 54-SpiralMatrix/54-SpiralMatrix.cpp
// Last updated: 5/17/2025, 3:49:00 PM
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();

        int lbound = 0;
        int rbound = matrix[0].size() - 1;
        int dbound = matrix.size() - 1;
        int ubound = 1;

        vector<int> res;

        int previ = 0;
        int prevj = -1;
        int nexti = 0;
        int nextj = 0;

        while (res.size() < m * n){
            res.push_back(matrix[nexti][nextj]);


            // from 3 to 6
                if ((nextj == rbound) && (previ == nexti) && (prevj < nextj)){
                    rbound -= 1;
                    prevj = nextj;

                    nexti++;
                }
                //from 9 to 8
                else if ((nexti == dbound) && (prevj == nextj) && (previ < nexti)){
                    dbound-= 1;
                    previ = nexti;

                    nextj--;
                }
                //from 7 to 4
                else if((nextj == lbound) && (previ == nexti) && (prevj > nextj)){
                    lbound += 1;
                    prevj =nextj;

                    nexti--;
                }
                //from 4 to 5
                else if ((nexti == ubound) && (prevj == nextj) && (previ > nexti)){
                    ubound +=1;
                    previ= nexti;

                    nextj++;
                }
                else if((prevj < nextj)){
                    prevj = nextj;
                    nextj++;

                }
                else if ((previ < nexti)){
                    previ = nexti;
                    nexti++;
                }
                else if (prevj > nextj){
                    prevj = nextj;
                    nextj--;
                }
                else if (previ > nexti){
                    previ = nexti;
                    nexti--;
                }

        }
        return res;
    }
};

File: 54-SpiralMatrix/54-SpiralMatrix.py
# Last updated: 7/2/2026, 11:17:44 PM
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:

        m = len(matrix)
        n = len(matrix[0])

        lbound = 0
        rbound = n - 1
        dbound = m - 1
        ubound = 1

        res: List[int] = []

        previ = 0
        prevj = -1
        nexti = 0
        nextj = 0

        while len(res) < m * n:
            res.append(matrix[nexti][nextj])

            # move right → down
            if (nextj == rbound) and (previ == nexti) and (prevj < nextj):
                rbound -= 1
                prevj = nextj
                nexti += 1

            # move down → left
            elif (nexti == dbound) and (prevj == nextj) and (previ < nexti):
                dbound -= 1
                previ = nexti
                nextj -= 1

            # move left → up
            elif (nextj == lbound) and (previ == nexti) and (prevj > nextj):
                lbound += 1
                prevj = nextj
                nexti -= 1

            # move up → right
            elif (nexti == ubound) and (prevj == nextj) and (previ > nexti):
                ubound += 1
                previ = nexti
                nextj += 1

            # continue right
            elif prevj < nextj:
                prevj = nextj
                nextj += 1

            # continue down
            elif previ < nexti:
                previ = nexti
                nexti += 1

            # continue left
            elif prevj > nextj:
                prevj = nextj
                nextj -= 1

            # continue up
            else:  # previ > nexti
                previ = nexti
                nexti -= 1

        return res